### ISSUE TITLE
[Operator] RoiAlign backend test case's output is misaligned

### ISSUE THREAD

**Issue by fdwr** [onnx/onnx#3428]
# Bug Report

### Is the issue related to model conversion?
No.

### Describe the bug
The [RoiAlign operator](https://github.com/onnx/onnx/blob/master/docs/Operators.md#RoiAlign), per the [Mask R-CNN paper](https://arxiv.org/abs/1703.06870) and Facebook Research's [Detectron 2 implementation](https://github.com/facebookresearch/detectron2/blob/master/detectron2/layers/roi_align.py) aligns sampling points over the center of the pixels, but the [solitary test case](https://github.com/onnx/onnx/blob/master/onnx/backend/test/case/node/roialign.py) in the ONNX backend has output that is misaligned by half an input pixel. After comparing to various references (table below under **Notes**) and reading more into the history of this operator, I see this is actually an old bug (which Detectron since fixed and PyTorch also fixed with aligned=True) which applied an offset the output subsample by 0.5 but forgot to adjust the input sample to compensate (see their comment in the [code](https://github.com/facebookresearch/detectron2/blob/master/detectron2/layers/roi_align.py): "_the original roi_align (aligned=False) does not subtract the 0.5 when computing neighboring pixel indices and therefore it uses pixels with a slightly incorrect alignment (relative to our pixel model) when performing bilinear interpolation_" and the [test case](https://github.com/facebookresearch/detectron2/blob/master/tests/layers/test_roi_align.py#L15)).

From the paper, note pixel centers used for interpolation:
![image](https://user-images.githubusercontent.com/1809166/110403224-95ce2500-8031-11eb-80c0-92732fa247d2.png)

We should fix the existing ONNX backend test case accordingly. Otherwise even a simple case of cropping an identity region from the input doesn't match. e.g. a new test case:

```python
    # Crop a 2x2 region from the 4x5 input.
    @staticmethod
    def export_roialign_identity_region():  # type: () -> None
        node = onnx.helper.make_node(
            "RoiAlign",
            inputs=["X", "rois", "batch_indices"],
            outputs=["Y"],
            spatial_scale=1.0,
            output_height=2,
            output_width=2,
            sampling_ratio=1,
        )

        X = np.array(
            [
                [
                    [
                        [ 0, 1, 2, 3],
                        [ 4, 5, 6, 7],
                        [ 8, 9,10,11],
                        [12,13,14,15],
                        [16,17,18,19],
                    ]
                ]
            ],
            dtype=np.float32,
        )
        batch_indices = np.array([0], dtype=np.int64)
        rois = np.array([[1, 2, 3, 4]], dtype=np.float32)
        # (num_rois, C, output_height, output_width)
        Y = np.array(
            [
                [
                    [
                        [ 9,10], # Should be identity scale/translation window into the input,
                        [13,14], # and not [[11.5,12.5], [15.5,16.5]].
                    ]
                ]
            ],
            dtype=np.float32,
        )

        expect(node, inputs=[X, rois, batch_indices], outputs=[Y], name="test_roialign_identity_region")
```
[roi_align_identity_region.zip](https://github.com/onnx/onnx/files/6322528/roi_align_identity_region.zip)

### System information
- OS Platform and Distribution: Windows 10 20H2 19042.867
- ONNX version:  1.8.1
- Python version: 3.9.1
- GCC/Compiler version: VS2019
- CMake version: NA
- Protobuf version: NA
- Visual Studio version (if applicable): VS2019

### Reproduction instructions
- Run the existing ONNX back test case.

### Expected behavior
The test case should bilinearly sample from pixel centers in *both* input and output pixels, not half of one and zero of the other.

e.g.
![image](https://user-images.githubusercontent.com/1809166/110191631-5d82d880-7dde-11eb-938b-da13a22eda97.png)

### Notes

I compared the bounding box output for the [Mask R-CNN model](https://github.com/onnx/models/tree/master/vision/object_detection_segmentation/mask-rcnn/model) using the baseline output in the tar.gz vs the half-pixel aligned output from the ONNX Runtime DirectML execution provider, and I feel the alignment is tighter and more accurately aligned, with fewer random false positives in the half-pixel sampled implementation. Other data scientists have also complained online that the old bug in PyTorch negatively affected their object detection results.

![image](https://user-images.githubusercontent.com/1809166/114975841-c75cbc00-9e39-11eb-8b7f-23e4ceb11d22.png)

### Framework comparisons

```
Input Tensor =
                   <------->
              0.0 1.0 2.0 3.0 4.0 5.0 6.0
               |.5 |.5 |.5 |.5 |.5 |.5 |
        0.0___ |_|_|_|_|_|_|_|_|_|_|_|_|
        1.0___[| 0,| 1,| 2,| 3,| 4,| 5 ]
    /|\ 2.0___[|10,┃11,┃12,┃13,|14,|15 ]
    \|/ 3.0___[|20,┃21,┃22,┃23,|24,|25 ]
        4.0___[|30,|31,|32,|33,|34,|35 ]
        5.0___[|40,|41,|42,|43,|44,|45 ]
        6.0___[|50,|51,|52,|53,|54,|55 ]

Active region of interest = [[1.0, 1.0, 3.0, 3.0], ...]
// The first window is 2x2 over the input elements, covering values [[11,12],[21,22]].
Output tensor size = [4,4]
```

Image | Source | Output 4x4, from first 2x2 region
-- | -- | --
![image](https://user-images.githubusercontent.com/1809166/110462790-fe46f180-8085-11eb-8af6-495337104ba2.png) | ✔ FB Research [Detectron 2](https://github.com/facebookresearch/detectron2/blob/master/detectron2/layers/roi_align.py) (MaskedRCNN paper) | [ 8.25,  8.75,  9.25,  9.75],<br/>[13.25, 13.75, 14.25, 14.75],<br/>[18.25, 18.75, 19.25, 19.75],<br/>[23.25, 23.75, 24.25, 24.75]
![image](https://user-images.githubusercontent.com/1809166/110462790-fe46f180-8085-11eb-8af6-495337104ba2.png) | ✔ ONNX Runtime DirectML EP (ROI_ALIGN 0) | [ 8.25,  8.75,  9.25,  9.75],<br/>[13.25, 13.75, 14.25, 14.75],<br/>[18.25, 18.75, 19.25, 19.75],<br/>[23.25, 23.75, 24.25, 24.75]
![image](https://user-images.githubusercontent.com/1809166/110462790-fe46f180-8085-11eb-8af6-495337104ba2.png) | ✔ ONNX Runtime 1.7 CPU [Resize](https://github.com/onnx/onnx/blob/master/docs/Operators.md#Resize) + Slice<br/>coordinate_transformation_mode=half | [ 8.25,  8.75,  9.25,  9.75],<br/>[13.25, 13.75, 14.25, 14.75],<br/>[18.25, 18.75, 19.25, 19.75],<br/>[23.25, 23.75, 24.25, 24.75]
![image](https://user-images.githubusercontent.com/1809166/110462790-fe46f180-8085-11eb-8af6-495337104ba2.png) | ✔ [`torchvision.ops.roi_align(aligned=True…)`](https://pytorch.org/vision/0.8/_modules/torchvision/ops/roi_align.html) | [ 8.25,  8.75,  9.25,  9.75],<br/>[13.25, 13.75, 14.25, 14.75],<br/>[18.25, 18.75, 19.25, 19.75],<br/>[23.25, 23.75, 24.25, 24.75]
![image](https://user-images.githubusercontent.com/1809166/110462870-174fa280-8086-11eb-8bf4-67a7603cb474.png) | ✖ [`torchvision.ops.roi_align(aligned=False…)`](https://pytorch.org/vision/0.8/_modules/torchvision/ops/roi_align.html)<br/>*deprecated, legacy flag still exists | [13.75, 14.25, 14.75, 15.25],<br/>[18.75, 19.25, 19.75, 20.25],<br/>[23.75, 24.25, 24.75, 25.25],<br/>[28.75, 29.25, 29.75, 30.25]
![image](https://user-images.githubusercontent.com/1809166/110462870-174fa280-8086-11eb-8bf4-67a7603cb474.png) | ✖ ONNX Runtime 1.7 CPU EP RoiAlign<br/>[Issue \#6921](https://github.com/microsoft/onnxruntime/issues/6921) | [13.75, 14.25, 14.75, 15.25],<br/>[18.75, 19.25, 19.75, 20.25],<br/>[23.75, 24.25, 24.75, 25.25],<br/>[28.75, 29.25, 29.75, 30.25]
![image](https://user-images.githubusercontent.com/1809166/110463016-3f3f0600-8086-11eb-8a37-da7159ae65cd.png) | 🤨 [`tf.image.crop_and_resize(…)`](https://www.tensorflow.org/api_docs/python/tf/image/crop_and_resize)<br/>*Note boxes are normalized 0 to 1 (so /5 each ROI element) | [11.00, 11.66, 12.33, 13.00],<br/>[17.66, 18.33, 19.00, 19.66],<br/>[24.33, 25.00, 25.66, 26.33],<br/>[31.00, 31.66, 32.33, 33.00]
![image](https://user-images.githubusercontent.com/1809166/110463016-3f3f0600-8086-11eb-8a37-da7159ae65cd.png) | 🤨 `tf.image.resize_bilinear(align_corners=True…)`<br/> + `tf.slice` | [11.00, 11.66, 12.33, 13.00],<br/>[17.66, 18.33, 19.00, 19.66],<br/>[24.33, 25.00, 25.66, 26.33],<br/>[31.00, 31.66, 32.33, 33.00]
![image](https://user-images.githubusercontent.com/1809166/110462954-30585380-8086-11eb-9548-82558a07d044.png) | 🤨 `tf.image.resize_bilinear(align_corners=False…)`<br/> + `tf.slice` | [11.00, 11.50, 12.00, 12.50],<br/>[16.00, 16.50, 17.00, 17.50],<br/>[21.00, 21.50, 22.00, 22.50],<br/>[26.00, 26.50, 27.00, 27.50]
![image](https://user-images.githubusercontent.com/1809166/110462790-fe46f180-8085-11eb-8af6-495337104ba2.png) | ✔ `tf.image.resize_bilinear(half_pixel_centers=True…)`<br/> + `tf.slice` | [ 8.25,  8.75,  9.25,  9.75],<br/>[13.25, 13.75, 14.25, 14.75],<br/>[18.25, 18.75, 19.25, 19.75],<br/>[23.25, 23.75, 24.25, 24.75]

----------

Facebook research's [Detectron 2 test code](https://github.com/facebookresearch/detectron2/blob/master/tests/layers/test_roi_align.py#L15):

```python
class ROIAlignTest(unittest.TestCase):
    def test_forward_output(self):
        input = np.arange(25).reshape(5, 5).astype("float32")
        """
        0  1  2   3 4
        5  6  7   8 9
        10 11 12 13 14
        15 16 17 18 19
        20 21 22 23 24
        """

        output = self._simple_roialign(input, [1, 1, 3, 3], (4, 4), aligned=False)
        output_correct = self._simple_roialign(input, [1, 1, 3, 3], (4, 4), aligned=True)

        # without correction:
        old_results = [
            [7.5, 8, 8.5, 9],
            [10, 10.5, 11, 11.5],
            [12.5, 13, 13.5, 14],
            [15, 15.5, 16, 16.5],
        ]

        # with 0.5 correction:
        correct_results = [
            [4.5, 5.0, 5.5, 6.0],
            [7.0, 7.5, 8.0, 8.5],
            [9.5, 10.0, 10.5, 11.0],
            [12.0, 12.5, 13.0, 13.5],
        ]
        # This is an upsampled version of [[6, 7], [11, 12]]
...
```
----------

PyTorch sample code:
```python
# pip install torch==1.7.1+cpu torchvision==0.8.2+cpu torchaudio===0.7.2 -f https://download.pytorch.org/whl/torch_stable.html
import torch
import torchvision
print("PyTorch version:", torch.__version__)

input = [[[[ 0, 1, 2, 3, 4, 5], # NCHW
            [10,11,12,13,14,15],
            [20,21,22,23,24,25],
            [30,31,32,33,34,35],
            [40,41,42,43,44,45],
            [50,51,52,53,54,55]]]]
boxes = [[0, 1,1,3,3]]
output_size = [4,4]
aligned=True # Correct
#aligned=False # Legacy setting
sampling_ratio=1
spatial_scale=1

# https://pytorch.org/vision/0.8/_modules/torchvision/ops/roi_align.html
output = torchvision.ops.roi_align(
    torch.tensor(input, dtype=torch.float),
    torch.tensor(boxes, dtype=torch.float),
    output_size,
    spatial_scale=spatial_scale,
    sampling_ratio=sampling_ratio,
    aligned=aligned
)

torch.set_printoptions(sci_mode=False)
print(input)
print(boxes)
print(output)
```

----------

TensorFlow sample code:
```python
# pip install tensorflow-gpu==1.15.0
import os
import tensorflow.compat.v1 as tf

input = [[ # NHWC
            [[ 0.], [ 1.], [ 2.], [ 3.], [ 4.], [ 5.]],
            [[10.], [11.], [12.], [13.], [14.], [15.]],
            [[20.], [21.], [22.], [23.], [24.], [25.]],
            [[30.], [31.], [32.], [33.], [34.], [35.]],
            [[40.], [41.], [42.], [43.], [44.], [45.]],
            [[50.], [51.], [52.], [53.], [54.], [55.]]
        ]]
boxes = [[1/5,1/5,3/5,3/5],[3/5,3/5,4/5,4/5]] # Normalized 0.0 to 1.0 (where 1.0 = width - 1 and height - 1)
box_indices = [0, 0] # Batch indices per corresponding region
crop_size = [4, 4] # Output tensor size HW

print("TensorFlow version:", tf.__version__) # 1.15.0 (cpu/cuda)

# Using half_pixel_centers=True is correct (not align_corners=True)
output_size = [6*2, 6*2]
resize_output = tf.image.resize_bilinear(tf.constant(input), output_size, align_corners=False, half_pixel_centers=True)
resize_bilinear_slice_output = tf.slice(resize_output, [0,2,2,0], [1,4,4,1])

# Note crop_and_resize doesn't scale the image boundaries to pixel centers, but always to corners,
# and there is sadly no flag to influence this (unlike resize_bilinear).
method = 'bilinear'
extrapolation_value = 0
crop_and_resize_output = tf.image.crop_and_resize(
    image=tf.constant(input, dtype=tf.float32), # NHWC
    boxes=tf.constant(boxes, dtype=tf.float32),
    box_ind=tf.constant(box_indices, dtype=tf.int32),
    crop_size=tf.constant(crop_size, dtype=tf.int32),
    method=method,
    extrapolation_value=extrapolation_value
)

with tf.Session(config=config) as session:
    with np.printoptions(precision=3, suppress=True):
        print("input:\n", input)
        print("crop_and_resize:\n", session.run(crop_and_resize_output))
        print("resize_bilinear_and_slice:\n", session.run(resize_bilinear_slice_output))
```

**Comment by fdwr** [comment]
@sampepose @BowenBao @pranavsharma

**Comment by jiafatom** [comment]
Need update onnx spec -- add the attribute `aligned` to ROIAlign.
